src/seq.rs

```diff
@@ -0,0 +1,309 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Functions for randomly accessing and sampling sequences.
+
+use super::Rng;
+use std::collections::hash_map::HashMap;
+
+/// Randomly sample *up to* `amount` elements from a finite iterator.
+///
+/// The values are non-repeating but the order of elements returned is *not* random.
+///
+/// This implementation uses `O(len(iterable))` time and `O(amount)` memory.
+///
+/// > If `len(iterable) <= amount` then the values will be in sequential order. In all other
+/// > cases the order of the elements is neither random nor guaranteed.
+///
+/// # Example
+///
+/// ```rust
+/// use rand::{thread_rng, seq};
+///
+/// let mut rng = thread_rng();
+/// let sample = seq::sample_iter(&mut rng, 1..100, 5);
+/// println!("{:?}", sample);
+/// ```
+pub fn sample_iter<T, I, R>(rng: &mut R, iterable: I, amount: usize) -> Vec<T>
+    where I: IntoIterator<Item=T>,
+          R: Rng,
+{
+    let mut iter = iterable.into_iter();
+    let mut reservoir = Vec::with_capacity(amount);
+    reservoir.extend(iter.by_ref().take(amount));
+
+    // continue unless the iterator was exhausted
+    //
+    // note: this prevents iterators that "restart" from causing problems.
+    // If the iterator stops once, then so do we.
+    if reservoir.len() == amount {
+        for (i, elem) in iter.enumerate() {
+            let k = rng.gen_range(0, i + 1 + amount);
+            if let Some(spot) = reservoir.get_mut(k) {
+                *spot = elem;
+            }
+        }
+    } else {
+        // Don't hang onto extra memory. There is a corner case where
+        // `amount <<< len(iterable)` that we want to avoid.
+        reservoir.shrink_to_fit();
```

Comment:
So you're not going with your own advice in #194 to make this fallible? I'm not sure personally; this is an *iterator* so shrinking to available data is common behaviour.